core/src/main/scala/io/projectglow/sql/expressions/Impute.scala

```diff
@@ -0,0 +1,171 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.projectglow.sql.expressions
+
+import io.projectglow.sql.util.RewriteAfterResolution
+
+import org.apache.spark.sql.SQLUtils
+import org.apache.spark.sql.catalyst.analysis.TypeCheckResult
+import org.apache.spark.sql.catalyst.dsl.expressions._
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.catalyst.expressions.aggregate.Average
+import org.apache.spark.sql.catalyst.expressions.codegen.{CodegenContext, ExprCode}
+import org.apache.spark.sql.catalyst.util.ArrayData
+import org.apache.spark.sql.types.{ArrayType, DataType, IntegerType, NumericType, StringType, StructType}
+import org.apache.spark.unsafe.types.UTF8String
+
+/**
+ * Imputes the missing values of an array using the non-missing values. Values that are NaN, null or equal to the
+ * missing value parameter are not included in the aggregation, and are substituted with the specified strategy.
+ * Currently, the supported strategies are mean and median. If all values are missing, they are substituted with the
+ * missing value.
+ *
+ * If the missing value is not provided, the parameter defaults to -1.
+ */
+case class Impute(array: Expression, strategy: Expression, missingValue: Expression)
+    extends RewriteAfterResolution {
+
+  override def children: Seq[Expression] = Seq(array, strategy, missingValue)
+
+  def this(array: Expression, strategy: Expression) = {
+    this(array, strategy, Literal(-1))
+  }
+
+  // Mapping from imputation strategy to the imputed value
+  lazy val imputationStrategies: Map[String, Expression] = Map(
+    "mean" -> getMean,
+    "median" -> getMedian
+  )
+
+  // A value is considered missing if it is NaN, null or equal to the missing value parameter
+  def isMissing(lv: NamedLambdaVariable): Predicate = IsNaN(lv) || IsNull(lv) || lv === missingValue
+
+  def getMean: Expression = {
+    val nLv =
+      NamedLambdaVariable("numArg", array.dataType.asInstanceOf[ArrayType].elementType, true)
+    val sLv = NamedLambdaVariable("structArg", StructType.fromDDL("sum double, count long"), true)
+    val sumName = Literal(UTF8String.fromString("sum"), StringType)
+    val countName = Literal(UTF8String.fromString("count"), StringType)
+    val getSum = GetStructField(sLv, 0, Some("sum"))
+    val getCount = GetStructField(sLv, 1, Some("count"))
+
+    // Sum and count of non-missing values
+    val zeros = namedStruct(sumName, Literal(0d), countName, Literal(0L))
+    // Update sum and count with array value
```

Comment:
We should use an algorithm that's not as likely to overflow, for example https://stackoverflow.com/a/1346890

Reply:
I'm not sure if we're concerned about overflow in the common case. Given that we're generally working with genotype states (0, 1, or 2) and supporting <10M individuals, I don't think we'd hit the maximum value of a Double (~1.8E308). Calculating the mean via this method will require (N-1) additional divisions and (potentially but less importantly) introduce increased floating point error.